R/contribution_matrix.R

```diff
@@ -0,0 +1,588 @@
+    effect_variance <- data.frame(Study = data_no_control$Study,
+                                  Treatment = data_no_control$Treatment,
+                                  Variance = effect_sizes$vi)
+    
+    #Calculate variance in the control arm
+    control_mean <- metafor::escalc(measure = "MN",
+                                    mi = data_control$Mean,
+                                    sdi = data_control$SD,
+                                    ni = data_control$N)
+    control_variance <- data.frame(Study = data_control$Study,
+                                   Treatment = data_control$Treatment,
+                                   Variance = control_mean$vi)
+  } else{
+    stop("outcome_type must be 'Continuous' or 'Binary'")
+  }
+  
+  #Create the V matrix with treatment effect variances on the leading diagonal
+  V <- diag(effect_variance$Variance)
+  rownames(V) <- study_treatment
+  colnames(V) <- study_treatment
+  #Number of arms per study
+  n_arms <- table(data_wide$Study)
+  #Insert covariances into multi-arm studies. The covariance is always the variance in the control arm.
+  for(i in 1:n_studies){
+    if(n_arms[studies[i]] > 1){
+      #Create a small matrix for this study only, with the control variance everywhere
+      mini_V <- matrix(control_variance$Variance[control_variance$Study == studies[i]],
+                       nrow = n_arms[studies[i]],
+                       ncol = n_arms[studies[i]])
+      #Put the treatment effect variances on the leading diagonal
+      diag(mini_V) <- effect_variance$Variance[effect_variance$Study == studies[i]]
+      #Overwrite the submatrix in V corresponding to this study
+      V[which(effect_variance$Study == studies[i]), which(effect_variance$Study == studies[i])] <- mini_V
+    }
+  }
+  return(V)
+}
+
+
+
+#' Create the design matrix.
+#' 
+#' @param data Input data in long format plus the column 'Treatment', a text version of 'T'.
+#' @param studies Vector of studies.
+#' @param treatments Vector of treatments with the reference treatment first.
+#' @param covar_centered Vector of centered covariate values, named by study.
+#' @param cov_parameters "Shared", "Exchangeable", or "Unrelated".
+#' @return X matrix.
+CreateXMatrix <- function(data, studies, treatments, covar_centered, cov_parameters){
+  #Number of studies
+  n_studies <- length(studies)
+  #Number of treatments
+  n_treatments <- length(treatments)
+  #Reference treatment
+  reference <- treatments[1]
+  #Data with control treatment rows deleted
+  data_no_control <- DeleteControlTreatment(data = data, treatments = treatments)
+  #Used as labels for the matrix rows
+  study_treatment <- paste0("(", data_no_control$Study, ")", data_no_control$Control, ":", data_no_control$Treatment)
+  
+  #Create the design matrix with the right dimensions and zeros everywhere
+  if(cov_parameters %in% c("Unrelated", "Exchangeable")){
+    X <- matrix(0, nrow = length(study_treatment), ncol = 2 * (n_treatments - 1))
```

Comment:
Single-letter variable names are discouraged because it's difficult to identify what they are meant to represent. It'd fine in a 1-line function, but here a more explicit name would be helpful

Reply:
Ok, not ideal, but makes sense here.